contracts/smart-account/modules/AccountRecoveryModule.sol

```diff
@@ -232,27 +244,77 @@ contract AccountRecoveryModule is
         bytes4 innerSelector;
         assembly {
             innerSelector := mload(add(innerCallData, 0x20))
         }
+
         bool isValidAddingRequestUserOp = (innerSelector ==
             this.submitRecoveryRequest.selector) &&
             (dest == address(this)) &&
             callValue == 0;
-        if (
-            isValidAddingRequestUserOp != //this a userOp to submit Recovery Request
-            (_smartAccountSettings[msg.sender].securityDelay == 0) //securityDelay is 0,
-        ) {
+        if (isValidAddingRequestUserOp) {
+            //check this is a request through SA.execute => this.executeRecovery
+            bytes4 expectedExecuteSelector;
+            address expectedThisAddress;
+            assembly {
+                //32(memory bytes array length) + 4(submitRecoveryRequest selector) + 32(offset) + 32(length)
+                expectedExecuteSelector := mload(add(innerCallData, 0x64))
+                expectedThisAddress := mload(add(innerCallData, 0x68))
+            }
+            if (
+                expectedExecuteSelector != EXECUTE_SELECTOR &&
+                expectedExecuteSelector != EXECUTE_OPTIMIZED_SELECTOR
+            ) {
+                revert("AccRecovery: WRR01"); //Wrong Recovery Request 01 = wrong execute selector in the request
+            }
+            if (expectedThisAddress != address(this)) {
+                revert("AccRecovery: WRR02"); //Wrong Recovery Request 02 = call should be to this contract
+            }
+            bytes4 expectedExecuteRecoverySelector;
+            uint256 executeRecoveryCallDataOffset;
+
+            assembly {
+                executeRecoveryCallDataOffset := mload(add(innerCallData, 0xa8))
+                expectedExecuteRecoverySelector := mload(
+                    add(
+                        innerCallData,
+                        add(
+                            //executeRecovery callData start position
+                            add(
+                                0x68, //position where execute() arguments start
+                                executeRecoveryCallDataOffset //offset where executeRecovery callData bytes array start
+                            ),
+                            //skip length
+                            0x20
+                        )
+                    )
+                )
+            }
+            if (
+                expectedExecuteRecoverySelector != this.executeRecovery.selector
+            ) {
+                revert("AccRecovery: WRR03"); //Wrong Recovery Request 03 = wrong executeRecovery selector in the request
+            }
+        }
+
+        bool isValidRecoveryExecutionUserOp = (innerSelector ==
+            this.executeRecovery.selector &&
+            dest == address(this) &&
+            _smartAccountSettings[smartAccount].securityDelay == 0);
+
+        if (isValidAddingRequestUserOp != isValidRecoveryExecutionUserOp) {
+            //exactly one should be true
```

Comment:
I think it would be beneficial to split this function into 2-3 subfunctions. Having ~200 lines for a function makes it hard to read, especially with all these conditions and assembly code.
Something like:
- checkRecoveryStatus
- validateSignatures
- verifyRecoveryRequest

Reply:
Thank you for inspiring me to do this :) Now validateUserOp looks much prettier and easier to understand